### PR TITLE
A new PR to resolve the conflicts in #78.

### PR DESCRIPTION
**Pull request by tel3path**
#78 could not be reopened, but has conflicts that need to be resolved. This new PR does not introduce any changes beyond what should have been merged from #78 and #79.

### CODE REVIEW

**Comment by tel3path** [comment]
bors try

**Comment by ltratt** [comment]
This has got 18 commits, so something has gone wrong.

**Comment by tel3path** [comment]
bors try

**Comment by 0152la** [comment]
Have there been any changes between this try and the previous one?

**Comment by tel3path** [comment]
> If not, I personally don't see a reason for the try; if you have one though, please tell me and let it continue. (also cancelling won't do anything on the buildbot side for now, unfortunately, so it'll continue either way)

The only reason is I thought my connection might have been dropped when I did the force push.

**Comment by 0152la** [comment]
I don't see a force push from you in this PR. Which one are you referring to?

**Comment by tel3path** [comment]
> 

That's exactly why I thought my connection was dropped. Because I don't see the force push either.

**Comment by tel3path** [comment]
> 

None of my attempts at force push are showing here. 
I did:
`git rebase -i upstream/master`
and got
`Successfully rebased and updated refs/heads/compare_platforms.`
then:
`git push origin +compare_platforms:master`
and got
`Everything up-to-date`

And since that didn't show I tried 
`git push origin compare_platforms -f`
and got
`Everything up-to-date`

**Comment by tel3path** [comment]
bors try

**Comment by tel3path** [comment]
bors try

**Comment by ltratt** [comment]
This is now starting to look about right. But before we actually merge this, I just realised that the C file is using tabs. Please convert all tabs to 4 spaces, please (as a new commit, not a force push).

**Comment by 0152la** [comment]
I thought indenting should be covered by `clang-format`. We could amend the options we use to check for formatting.

**Comment by ltratt** [comment]
It may or may not be fussy about tabs. Dunno. Bitter experience has taught me to be fussy about tabs, however!

**Comment by tel3path** [comment]
> 

OK. I promise you my editor is using spaces and not tabs. Will convert now.

**Comment by 0152la** [comment]
Sure, what I'm saying is `clang-format` might be forcing tabs. And I think it is, perusing the repo. If we want spaces, we'd need to amend the `clang-format` check appropriately in the CI script.

**Comment by ltratt** [comment]
Erp, OK, I stand correcte!. We should fix the clang-format settings, but not for this PR. My apologies -- please ignore what I said!

**Comment by tel3path** [comment]
> Erp, OK, I stand correcte!. We should fix the clang-format settings, but not for this PR. My apologies -- please ignore what I said!

OK

**Comment by tel3path** [comment]
> 

This, below:

> buildbot/ci-builder Expected — Waiting for status to be reported

I haven't seen this before. 

It also says

> Required statuses must pass before merging
All required [statuses](https://docs.github.com/rest/commits/statuses) and check runs on this pull request must run successfully to enable automatic merging.

Is this a step that bors will complete, or is it something we need to fix?

**Comment by ltratt** [comment]
@0152la I think this now looks in good order so will leave it to you to shepherd through.

**Comment by tel3path** [comment]
> This is now starting to look about right. But before we actually merge this, I just realised that the C file is using tabs. Please convert all tabs to 4 spaces, please (as a new commit, not a force push).

> 

This, below:

> buildbot/ci-builder Expected — Waiting for status to be reported

I haven't seen this before. 

It also says

> Required statuses must pass before merging
All required [statuses](https://docs.github.com/rest/commits/statuses) and check runs on this pull request must run successfully to enable automatic merging.

Is this a step that bors will complete, or is it something we need to fix?

**Comment by 0152la** [comment]
Ignore those for now; we've made some repo changes, and I added some stuff which I'm not entirely sure what will happen, but at least it will hopefully prevent what happened with the previous PR.

File: compare_platforms/compare_platforms_read_only.c
Comment (0152la) [inline compare_platforms/compare_platforms_read_only.c:18]: It seems this formatting got broken, perhaps due to the `clang-format` call. Can you please try to find a way to redo it?

**Review by 0152la — CHANGES_REQUESTED** [review]
In addition to the extra comment, it seems a binary `compare_platforms_read_only` snuck in. Please delete it.

File: compare_platforms/compare_platforms_read_only.c
Comment (tel3path) [reply compare_platforms/compare_platforms_read_only.c:17]: Binary fixed in 35d8ec4d941d9833b54920d07b9b530f76f97afe

Top comment formatting fixed (and content corrected to reflect a change to the program flow) in 738e5ca2d1b374e09fb29b537cd28a75aded7e3e

**Review by 0152la — APPROVED** [review]

**Comment by 0152la** [comment]
Please squash.

**Comment by tel3path** [comment]
> 

Top comment says "can_read" when it should say "can_write". Will correct first.

**Comment by tel3path** [comment]
> Please squash.

Squashed.

**Comment by 0152la** [comment]
Ah, you squashed before I had a chance to look at the new commit; should have pinged me first, and then I would've told you to squash again. Am I correct the now missing commit only changed one line [1]?

[1] https://github.com/capablevms/cheri-examples/pull/80/files#diff-dc9ca6bc2cc4144f4403159f06c492ad90e9c962dbf118111e2603f2145448c7R13

**Comment by tel3path** [comment]
> Ah, you squashed before I had a chance to look at the new commit; should have pinged me first, and then I would've told you to squash again. Am I correct the now missing commit only changed one line [1]?
> 
> [1] https://github.com/capablevms/cheri-examples/pull/80/files#diff-dc9ca6bc2cc4144f4403159f06c492ad90e9c962dbf118111e2603f2145448c7R13

Yes. Sorry about that. The only change was to that one line.

**Comment by tel3path** [comment]
> Alright. Then I'll ask you to amend the message of the commit so that it summarizes the new test we add; no need to keep any of the detail we added during the PR process.

Fixed in 5c9aa347f4d332ec689e0190be5a8f40a9fa1a2d

**Comment by 0152la** [comment]
I still see the commit title being `parent 150aeec`, and some additional meta data at the top. Can we please remove these. For the body of the commit, can you be a bit more descriptive with what the test does; do feel free to lift things from the preamble of the test itself. 

You mention `morello-purecap`. Our CI should run the test on RISC-V as well, and I believe we would expect this test to be compile-able for non-CHERI platforms. Is there any reason you explicitly specified "on morello-purecap"?

Finally, I believe the changed tests you mentioned were done in #79, not in this PR.

**Comment by tel3path** [comment]
> I believe we would expect this test to be compile-able for non-CHERI platforms. 

Unless I'm missing something, run_tests.sh recognizes riscv64, morello-purecap or morello-hybrid only; and this particular test would be run in only the first two of those three environments.

**Comment by 0152la** [comment]
Yes, that is correct; our CI only runs these tests on the three CHERI configurations you named. However, there is nothing in the test itself (like an `ifndef (CHERI); error;`) to prevent someone from compiling for another configuration. And indeed, the test is meant to be run in a non-CHERI environment, to expose the different functionality. As such, we should be careful when specifying "this test is made for this configuration". There are legitimate reasons to restrict this (e.g., using Morello only ASM instructions), but in this case, there is no such limitation.

**Comment by tel3path** [comment]
> 

I see what you mean. Currently, the script expects $BUILDBOT_PLATFORM to be riscv64, morello-purecap, or morello-hybrid, and throws an error if that environment variable is set to anything else. But there's no reason why it *couldn't* be extended to recognize other environment variables.

**Comment by 0152la** [comment]
Is there something within the test preventing it from being compiled with a non-CHERI aware compiler?

**Comment by tel3path** [comment]
> Is there something within the test preventing it from being compiled with a non-CHERI aware compiler?



> Is there something within the test preventing it from being compiled with a non-CHERI aware compiler?

No, as you pointed out above.

**Comment by 0152la** [comment]
Alright; let me know if there is further confusion regarding this point. We can take it on Mattermost as well to not clutter this space.

**Comment by 0152la** [comment]
I apologize, I'll have to ask you to rewrite it. I want a description about just the new test, `compare_platforms_read_only.c`,  briefly describing its functionality: we have a struct, we set its members read only, we try to overwrite them -- something along those lines (as I said, something similar to the preamble comment). 

The fact that we added it to our CI via `run_tests.sh` and how that script works is aside the main purpose of this PR, which is adding the new test.

**Comment by tel3path** [comment]
> I apologize, I'll have to ask you to rewrite it. I want a description about just the new test, `compare_platforms_read_only.c`, briefly describing its functionality: we have a struct, we set its members read only, we try to overwrite them -- something along those lines (as I said, something similar to the preamble comment).
> 
> The fact that we added it to our CI via `run_tests.sh` and how that script works is aside the main purpose of this PR, which is adding the new test.

Fixed in 8319754c06b8a62611a9725af12520db7b14cabc

**Comment by tel3path** [comment]
> Build failed:
> 
> * [buildbot/ci-builder](http://bencher14.soft-dev.org:8010/#builders/2/builds/102)

I will run clang-format on the c file again

**Comment by tel3path** [comment]
bors try

**Comment by 0152la** [comment]
Please squash.

**Comment by tel3path** [comment]
> Please squash.

Squashed.

**Pull request merged**
